Improve handling of module with numbers in title

gkaragia opened the pull request:
There was an issue with the 'Module slugs mismatch' tool in case the module name had numbers in the beggining. In that case `$sanitized_title` variable would begin with a number which would get appended to the backreference of preg_replace and cause weird behaviour.

### Changes proposed in this Pull Request

* Fixes the above issue by switching to `preg_match`. We could use curly braces and keep `preg_replace` but I think that using `preg_match` is a bit easier to understand.
* The number in the beginning of the slug is not prepended if the title begins with the same number. This is to avoid infinitely adding any starting numbers.
* There are still cases that the tool does not work correctly but I don't think that we can make it 100% accurate as long as we store the teacher id in the module slug and allow the users to modify the slug freely.

### Testing instructions

* Create a module with a name which starts with numbers.
* Run the tool and observe that the module slug is not modified.
* Manually modify the module slug and rerun the tool.
* Observe that the module slug is updated to the module title correctly.
* Create a module which is owned by a teacher and repeat the above steps.

donnapep commented:
Did a bit of testing. Here's my before snapshot:
![Screen Shot 2021-03-26 at 9 27 06 AM](https://user-images.githubusercontent.com/1190420/112638280-76970c00-8e15-11eb-9019-dc1b29ca9415.jpg)

Here's the after snapshot:
![Screen Shot 2021-03-26 at 9 27 38 AM](https://user-images.githubusercontent.com/1190420/112638331-86165500-8e15-11eb-9dad-ea9dcaec02eb.jpg)

On `master` branch, it was changing the module slug for the 123 Module, but wasn't touching the slugs of those other modules.

On the branch of this PR, it's changing the 2020 modules and the 30 module. Is that expected behavior?

gkaragia commented:
> On the branch of this PR, the slug for 123 Module is no longer change, but now it's changing the 2020 modules and the 30 module. Is that expected behavior?

Hey @donnapep yes. The correct slug for every module is the return value of `sanitize_title` with the the teacher id prepended if the teacher of the module is not an admin.

Things get tricky when the title of the module starts with a number as it is hard to tell by looking at the slug what is hapenning as there are many cases:
- The number is a teacher id 
- The number is part of the module title
- The number is a user modification of the slug

Previously there was a bug in the tool that the slug had its first number removed if the title started with a number which is now fixed.

In your example the `123 Module` has a correct slug so it should not be modified by the tool. `2020 Module 1, 2, & 3` are presumably owned by teacher with id 20 (as 020 does not exist in the beginning of the title). The module title for this cases does not exist in the slug though so the slug is replaced by `020 . <The module title>`. The last case is the same. `0` is assumed to be a teacher id (although it is obvious that it isn't) and the module title is appended after that.

We could possibly check to see if a slug begins with a `0` when repairing and don't regard it as a teacher id. But I think that we shouldn't dwell on this further because:
- As long as the user is able to freely modify the slug, it's not possible to create a tool that will work in every case.
- If the user does not modify the slug manually, there is no case that we would end up with a slug beggining with a `0`. This happened in your case because of the bug that existed and removed the first number of the title.

Sorry for the long answer, this whole thing is confusing 🙂

donnapep commented:
Closing as it's quite old.
